pkg/state/hcl_loader.go

```diff
@@ -0,0 +1,358 @@
+		},
+	}
+	// make sure content has a struct with helmfile_vars Schema defined
+	content, diags := hclFile.Body.Content(HelmfileHCLValuesSchema)
+	if diags != nil {
+		return nil, nil, diags
+	}
+
+	var helmfileLocalsVars map[string]*HelmfileHCLValue
+	// Decode blocks to return HelmfileHCLValue object => (each var with expr + Name )
+
+	if len(content.Blocks.OfType(localsBlockIdentifier)) > 1 {
+		return nil, nil, hcl.Diagnostics{
+			&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "A file can only support exactly 1 `locals` block",
+				Subject:  &content.Blocks[0].DefRange,
+			}}
+	}
+	for _, block := range content.Blocks {
+		var helmfileBlockVars map[string]*HelmfileHCLValue
+		if block.Type == valuesBlockIdentifier {
+			helmfileBlockVars, diags = hl.decodeHelmfileHCLValuesBlock(block)
+			if diags != nil {
+				return nil, nil, diags
+			}
+		}
+
+		if block.Type == localsBlockIdentifier {
+			helmfileLocalsVars, diags = hl.decodeHelmfileHCLValuesBlock(block)
+			if diags != nil {
+				return nil, nil, diags
+			}
+		}
+
+		// make sure vars are unique across blocks
+		for k := range helmfileBlockVars {
+			if hvars[k] != nil {
+				var diags hcl.Diagnostics
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Duplicate helmfile_vars definition",
+					Detail: fmt.Sprintf("The helmfile_var %q was already defined at %s:%d",
+						k, hvars[k].Range.Filename, hvars[k].Range.Start.Line),
+					Subject: &helmfileBlockVars[k].Range,
+				})
+				return nil, nil, diags
+			}
+		}
+		err = mergo.Merge(&hvars, &helmfileBlockVars)
+		if err != nil {
+			var diags hcl.Diagnostics
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Merge failed",
+				Detail:   err.Error(),
+				Subject:  nil,
+			})
+			return nil, nil, diags
+		}
+	}
+
+	return hvars, helmfileLocalsVars, nil
+}
+
+func (hl *HCLLoader) decodeHelmfileHCLValuesBlock(block *hcl.Block) (map[string]*HelmfileHCLValue, hcl.Diagnostics) {
+	attrs, diags := block.Body.JustAttributes()
+	if len(attrs) == 0 || diags != nil {
+		return nil, diags
+	}
+
+	hfVars := map[string]*HelmfileHCLValue{}
+	for name, attr := range attrs {
+		if !hclsyntax.ValidIdentifier(name) {
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid helmfile_vars variable name",
+				Detail:   badIdentifierDetail,
+				Subject:  &attr.NameRange,
+			})
+		}
+
+		hfVars[name] = &HelmfileHCLValue{
+			Name:  name,
+			Expr:  attr.Expr,
+			Range: attr.Range,
+		}
+	}
+	return hfVars, diags
+}
+
+func (hl *HCLLoader) parseSingleAttrRef(traversal hcl.Traversal, blockType string) (string, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+
+	root := traversal.RootName()
+	// In `values` blocks, Locals are always precomputed, so they don't need to be in the graph
+	if root == localsAccessorPrefix && blockType != localsBlockIdentifier {
+		return "", nil
+	}
+	rootRange := traversal[0].SourceRange()
```

Comment:
is ti safe? traversal[0]

Reply:
In this specific context, yes, pretty much (the function is not exported)
A traversal will always at least have 1 _traverser_
Example : 
`myvar = local.var`
The _traversal_ "local.var" object for myvar contains 2 _traversers_ : `local` and `var`.
The HCL parser would fail before if a variable was set like this 
`myvar = `
if we imagine than it would be interpreted as a nil traverser, which it is not.

There is one way that'd make it not safe though that is crossing my mind now : if a future contributor would purposely pass an empty traversal to the function (even if it'd not make much sense tbh). I'll add something to avoid that. Thank you